gradio/chat_interface.py

```diff
@@ -0,0 +1,269 @@
+                api_name=False,
+                queue=False,
+            ).then(
+                submit_fn,
+                [self.saved_input, self.chatbot],
+                [self.chatbot],
+                api_name=False,
+            )
+        
+        if self.retry_btn:
+            self.retry_btn.click(
+                self._delete_prev_fn,
+                [self.chatbot],
+                [self.chatbot, self.saved_input],
+                api_name=False,
+                queue=False,
+            ).then(
+                self._display_input,
+                [self.saved_input, self.chatbot],
+                [self.chatbot],
+                api_name=False,
+                queue=False,
+            ).then(
+                submit_fn,
+                [self.saved_input, self.chatbot],
+                [self.chatbot],
+                api_name=False,
+            )
+
+        if self.delete_last_btn:
+            self.delete_last_btn.click(
+                self._delete_prev_fn,
+                [self.chatbot],
+                [self.chatbot, self.saved_input],
+                api_name=False,
+                queue=False,
+            )
+
+        if self.clear_btn:
+            self.clear_btn.click(
+                lambda :([], None), 
+                None, 
+                [self.chatbot, self.saved_input],
+                queue=False,
+                api_name=False)
+            
+    def _setup_api(self):
+        if inspect.isgeneratorfunction(self.fn):
+            api_fn = self._api_stream_fn
+        else:
+            api_fn = self._api_submit_fn
+
+        # Use a gr.State() instead of self.chatbot so that the API doesn't require passing forth
+        # a chat history, instead it is just stored internally in the state.
+        history = State([])
+
+        self.fake_api_btn.click(
+            api_fn,
+            [self.textbox, history],
+            [self.textbox, history],
+            api_name="chat",
+        )
+
+    def _clear_and_save_textbox(self, message):
+        return "", message
+
+    def _display_input(self, message: str, history: list[list[str]]):
+        history.append((message, None))
+        return history
+
+    def _submit_fn(self, message: str, history: list[list[str]]):
+        response = self.fn(message, history)
+        history[-1][1] = response
+        return history
+
+    def _stream_fn(self, message: str, history: list[list[str]]):
+        for response in self.fn(message, history):
+            history[-1][1] = response
+            yield history
+
+    def _api_submit_fn(self, message: str, history: list[list[str]]):
+        response = self.fn(message, history)
+        history.append([message, response])
+        return response, history
+        
+    def _api_stream_fn(self, message: str, history: list[list[str]]):
+        history = self._display_input(message, history)
+        for response in self.fn(message, history):
+            history[-1][1] = response
```

Comment:
history can actually be a `List[Tuple[str]]` which will error out when trying to do the item assignment. This is caused by the client since it converts everything to tuples (I think we should remove in 4.0). 


Repro


<img width="657" alt="image" src="https://github.com/gradio-app/gradio/assets/41651716/b3c67570-22fa-493d-ab0c-830dbc4c6cfa">

```bash
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.10/site-packages/gradio/routes.py", line 439, in run_predict
    output = await app.get_blocks().process_api(
  File "/home/user/.local/lib/python3.10/site-packages/gradio/blocks.py", line 1384, in process_api
    result = await self.call_function(
  File "/home/user/.local/lib/python3.10/site-packages/gradio/blocks.py", line 1103, in call_function
    prediction = await utils.async_iteration(iterator)
  File "/home/user/.local/lib/python3.10/site-packages/gradio/utils.py", line 343, in async_iteration
    return await iterator.__anext__()
  File "/home/user/.local/lib/python3.10/site-packages/gradio/utils.py", line 336, in __anext__
    return await anyio.to_thread.run_sync(
  File "/home/user/.local/lib/python3.10/site-packages/anyio/to_thread.py", line 33, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/home/user/.local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 877, in run_sync_in_worker_thread
    return await future
  File "/home/user/.local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 807, in run
    result = context.run(func, *args)
  File "/home/user/.local/lib/python3.10/site-packages/gradio/utils.py", line 319, in run_sync_iterator_async
    return next(iterator)
  File "/home/user/.local/lib/python3.10/site-packages/gradio/utils.py", line 688, in gen_wrapper
    yield from f(*args, **kwargs)
  File "/home/user/.local/lib/python3.10/site-packages/gradio/chat_interface.py", line 258, in _api_stream_fn
    history[-1][1] = response
TypeError: 'tuple' object does not support item assignment
```

Reply:
Nice catch